Convert bash scripts to Python

lilithxxx opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Convert bash scripts to Python
## Checklist
- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python -m scripts.pre_commit_linter` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR has an appropriate "PROJECT: ..." label (Please add this label for the first-pass review of the PR).
- [x] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [x] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

Commit:
initial

Commit:
fix lint

Commit:
fix

Commit:
fix

Commit:
conver

Commit:
fix

Commit:
fix

Commit:
convert

Commit:
fix

Commit:
Merge branch 'develop' into convert_bash_1

Commit:
Merge branch 'develop' into convert_bash_1

Commit:
add

Commit:
convert

Commit:
fix lint

Commit:
convert

Commit:
convert

Commit:
Merge branch 'develop' into convert_bash_1

Commit:
convert

Commit:
convert

Commit:
convert

Commit:
address

Commit:
fix path

Commit:
convert

Commit:
Merge branch 'develop' into convert_bash_1

Commit:
remove files

Commit:
convert

Commit:
fix comment

Commit:
convert

Commit:
convert

Commit:
fix

Commit:
address comments

Commit:
convert

Commit:
convert

Commit:
fix lint

Commit:
update circleci

Commit:
fix

Commit:
convert

Commit:
fix

Commit:
fix

Commit:
fix

Commit:
fix

Commit:
Merge branch 'develop' into convert_bash_1

Commit:
sync with develop

Commit:
fix

seanlip commented on scripts/backend_tests.py:
```diff
@@ -455,4 +448,4 @@ def main():
 
 if __name__ == '__main__':
-    main()
+    main(sys.argv)
```
Interesting -- can you explain the rationale behind this and similar changes elsewhere? The answer [here](https://stackoverflow.com/questions/35365344/python-sys-argv-and-argparse) suggests that argparse is preferable.

brianrodri replied:
Ah I suggested this: `argparse` checks `sys.argv` by default. This is just feeding it in explicitly.

Doing it this way allows us to test `main` by passing custom values of argv later (should we choose to test), and also invoke different scripts without needing to start a sub process.

seanlip replied:
Got it. I'm fine with it, but let's please add a comment at all locations of this saying what you said, because it's different from what seems to be the community norm.

Also, if you look at the SO link, we lose some benefits of argparse this way, e.g. the help function when the script is being run standalone. Has this been taken into account? Is there any way to keep those benefits?

brianrodri replied:
Sorry I'm confused, which functionality are we losing? These scripts are still using `argparse`, we're just explicitly passing `sys.argv` instead of allowing the module to look it up on its own ([doc](https://docs.python.org/2.7/library/argparse.html#beyond-sys-argv)). See line 310: the only difference is how it eventually gets its hands on `sys.argv`.

seanlip replied:
Oh, sorry! I think I read this wrongly and thought it was a replacement for argparse. My bad -- please ignore this comment.

(Also, just to check -- with the changes, the automatic argparse help message still does show when the script is run independently from the command line, right?)

seanlip commented on scripts/create_expression_parser.py:
```diff
@@ -0,0 +1,73 @@
+            '%s/bin/npm install pegjs@0.8.0' % common.NODE_PATH).split())
+
+    subprocess.call((
+        'node_modules/pegjs/bin/pegjs %s %s'
```
Should we use os.path.join for this, to accommodate other types of systems? (Ditto above and elsewhere.) 

In general I think we should be passing lists of the "command parts" into these types of functions.

lilithxxx replied:
Done

seanlip commented on scripts/create_expression_parser.py:
```diff
@@ -0,0 +1,73 @@
+    if not os.path.exists('node_modules/pegjs'):
+        python_utils.PRINT('Installing pegjs')
+        subprocess.call((
+            '%s/bin/npm install pegjs@0.8.0' % common.NODE_PATH).split())
```
Can you make a standard utility function in one of the python scripts that handles installation of npm libs, similar to what we did in bash? Have other things that need similar functionality make use of that utility function too.

Ideally you can generalize it to cover the whole block from lines 47-52.

lilithxxx replied:
Done

seanlip commented on scripts/create_expression_parser.py:
```diff
@@ -0,0 +1,73 @@
+        % (expression_parser_definition, expression_parser_js)).split())
+
+    for line in fileinput.input(files=[expression_parser_js], inplace=True):
+        python_utils.PRINT(
```
Why do you need to print the results of these commands?

lilithxxx replied:
Inside this loop the STDOUT will be redirected to the file.

seanlip replied:
Add a comment explaining.

lilithxxx replied:
I think I did

seanlip replied:
Oh yes, you did, sorry, I missed that. Apologies!

seanlip commented on scripts/create_expression_parser.py:
```diff
@@ -0,0 +1,73 @@
+                line), end='')
+
+        python_utils.PRINT(
+            re.sub(r'^})();\s*$', '}]);', line), end='')
```
Somewhere here I thought you would need to update the original file. I'm not sure where this is happening. Did you try deleting the generated file, running the script, and seeing if it produces the generated file?

lilithxxx replied:
Yes I manually checked it. I added the inplace arg

seanlip commented on scripts/install_chrome_on_travis.py:
```diff
@@ -0,0 +1,52 @@
+        os.chdir(oppia_dir)
+
+    python_utils.PRINT('Installing %s' % travis_chrome_path)
+    subprocess.call(('sudo dpkg -i %s' % travis_chrome_path).split())
```
Pass a list of args from the get-go. Ditto elsewhere.

lilithxxx replied:
Done

seanlip commented on scripts/install_third_party_libs.py:
```diff
@@ -0,0 +1,253 @@
+            'page:')
+
+        os_info = os.uname()
+        if os_info[0] != 'Darwin':
```
This is quite weird. Why "!="? Ditto below.

lilithxxx replied:
Sorry I meant ==. Changed

seanlip commented on scripts/install_third_party_libs.py:
```diff
@@ -0,0 +1,253 @@
+                'Windows%29')
+        sys.exit(1)
+
+    if hasattr(pip, 'main'):
```
Please add a comment to explain what is going on here.

lilithxxx replied:
Done

seanlip commented on scripts/install_third_party_libs.py:
```diff
@@ -0,0 +1,253 @@
+    """Download and install Skulpt. Skulpt is built using a Python script
+    included within the Skulpt repository (skulpt.py). This script normally
+    requires GitPython, however the patches to it below
+    (with the sed operations) lead to it no longer being required. The Python
```
What sed operations?

lilithxxx replied:
Removed

seanlip commented on scripts/install_third_party_libs.py:
```diff
@@ -0,0 +1,253 @@
+    ]
+
+    for package, version, path in pip_dependencies:
+        python_utils.PRINT(
```
Extract lines 220-226 into an ensure_pip_library_is_installed() function.

lilithxxx replied:
Done

seanlip commented on scripts/install_third_party_libs.py:
```diff
@@ -0,0 +1,253 @@
+    subprocess.call(('%s/bin/npm dedupe' % common.NODE_PATH).split())
+
+    install_skulpt(argv)
+    # Install pre-commit script.
```
Add newline above this one.

lilithxxx replied:
Done

seanlip commented on scripts/vagrant_lock.py:
```diff
@@ -0,0 +1,65 @@
+        sys.exit(0)
+
+    if os.path.isfile(vagrant_lock_file):
+        python_utils.PRINT('')
```
Can you create a utility function that takes in as input a list of strings, and prints each string on a separate line? I think that would read more cleanly than this, in many cases.

lilithxxx replied:
Done

seanlip reviewed (changes requested):
Thanks @lilithxxx, took a first pass. 

One question: did you test by manually running each script individually to make sure that it does the correct thing? If so, please list the "smoking gun" check for each script. E.g. running start.sh should result in a browser window opening with the Oppia dev server.

Commit:
fix conflicts

Commit:
use parse known args

Commit:
remove user

brianrodri commented on scripts/build.py:
```diff
@@ -1329,22 +1330,24 @@ def compile_typescript_files_continuously(project_dir):
 
 
-def build():
+def build(argv):
```
This should be called `main`.

lilithxxx replied:
Done

brianrodri commented on .travis.yml:
```diff
@@ -101,28 +103,28 @@ script:
-- if [ "$RUN_E2E_TESTS_TOPICS_AND_SKILLS_DASHBOARD" == 'true' ]; then travis_retry bash scripts/run_e2e_tests.sh --suite="topicsAndSkillsDashboard" --prod_env; fi
-- if [ "$RUN_E2E_TESTS_TOPIC_AND_STORY_EDITOR" == 'true' ]; then travis_retry bash scripts/run_e2e_tests.sh --suite="topicAndStoryEditor" --prod_env; fi
-- if [ "$RUN_E2E_TESTS_USERS" == 'true' ]; then travis_retry bash scripts/run_e2e_tests.sh --suite="users" --prod_env; fi
+- if [ "$RUN_E2E_TESTS_ACCESSIBILITY" == 'true' ]; then travis_retry python -m scripts.run_e2e_tests --suite="accessibility" --prod_env; fi
```
Can we move this selection logic into Python? It could be significantly cleaner there:

```python
E2E_SUITES = {
  'RUN_E2E_TESTS_ACCESSIBILITY': 'accessibility',
  # ...
}

for flag_name, suite_name in E2E_SUITES.items():
  if parser[flag_name]:
    run_e2e_tests.main('--suite="%s" --prod_env' % suite_name)
```

lilithxxx replied:
Actually the default value for the "suite" flag is "full" so if we just do python -m scripts.run_e2e_tests then it should run all the tests by default. I wonder why we don't do this here. Should I change this then?

/cc @seanlip @kevinlee12

seanlip replied:
Let's leave this suggested change for a future PR + discussion. We are splitting it up here so that Travis can parallelize, otherwise builds will take a very long time.

brianrodri commented on scripts/common.py:
```diff
@@ -16,11 +16,29 @@
 
 import python_utils
 
+_PARENT_DIR = os.path.abspath(os.path.join(os.getcwd(), os.pardir))
```
Avoid prefixing with `_` to stay consistent with other variables.

brianrodri commented on scripts/common.py:
```diff
@@ -164,6 +182,49 @@ def ensure_release_scripts_folder_exists_and_is_up_to_date():
+    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    result = sock.connect_ex(('localhost', port))
+    sock.close()
+    return bool(result)
```
Please use a contextmanager here since sockets are prone to failure:

```python
import contextlib

# ...

def is_port_open(port):
    with contextlib.closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as s:
        return bool(s.connect_ex(('localhost', port)))

lilithxxx replied:
Done

brianrodri commented on scripts/build.py:
```diff
@@ -1329,22 +1330,24 @@ def compile_typescript_files_continuously(project_dir):
 
 
-def build():
+def build(argv):
```
Please make `argv=None` in all definitions of `main` so that `module.main()` can be a valid call.

lilithxxx replied:
Done

brianrodri commented on scripts/run_backend_tests.py:
```diff
@@ -0,0 +1,125 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""INSTRUCTIONS:
```
Please move these instructions into the argparse help documentation. The docstring should simply be a one sentence description since we can now get the helpful information by running the script.

lilithxxx replied:
Done

brianrodri commented on scripts/run_backend_tests.py:
```diff
@@ -0,0 +1,125 @@
+
+  --verbose or -v
+
+IMPORTANT: Only one of --test_path and --test_target should be specified.
```
This is enforcable in argparse: https://docs.python.org/2/library/argparse.html#mutual-exclusion

lilithxxx replied:
Done

brianrodri commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,238 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""INSTRUCTIONS:
```
Same, instructions should now be kept in the argparse description: https://docs.python.org/2/library/argparse.html#mutual-exclusion.

lilithxxx replied:
Done

brianrodri commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,238 @@
+         % (common.GOOGLE_APP_ENGINE_HOME, app_yaml_filepath)).split()))
+
+    # Wait for the servers to come up.
+    while common.is_port_open(4444) or common.is_port_open(9001):
```
If we're waiting for the server to come up, why are we looping on the assumption that they have already been running?

Shouldn't we be looping on "while NOT is_port_open(): sleep()" ?

lilithxxx replied:
Changed name to is_port_closed

brianrodri commented on scripts/run_performance_tests.py:
```diff
@@ -0,0 +1,158 @@
+
+    if not common.is_port_open(8181):
+        python_utils.PRINT(
+            'There is already a server running on localhost:8181')
```
We're checking that the port isn't open, and then complaining that it is open? Isn't this a typo?

lilithxxx replied:
Ditto

seanlip commented:
Please note that all the automated checks seem to be failing.

Commit:
fix pip

Commit:
try subprocess on pip install

Commit:
fix

Commit:
fix

Commit:
fix

Commit:
fix

Commit:
fix

Commit:
add node path

Commit:
fix

Commit:
fix lint

Commit:
add args

Commit:
fix lint

Commit:
merge develo

Commit:
add path

Commit:
address comments

Commit:
address comments

Commit:
fix

lilithxxx commented:
@seanlip @brianrodri PTAL

Commit:
fix

Commit:
fix

Commit:
fix

seanlip commented:
@lilithxxx The tests are still failing. Please re-ping when they pass.

Commit:
fix frontend

Commit:
fix

Commit:
fix start

Commit:
fix lint

Commit:
clear cache

Commit:
revert

Commit:
remove pylint

Commit:
remove pylint path

Commit:
add path

Commit:
add pylint path

Commit:
fix

Commit:
fix

Commit:
use vurtual env

Commit:
fix

Commit:
fix

Commit:
add url in travis yaml

Commit:
test

Commit:
test

Commit:
fix

Commit:
test

Commit:
test

Commit:
test

Commit:
test

Commit:
test

brianrodri commented on .circleci/config.yml:
```diff
@@ -15,6 +15,7 @@ anchor_for_job_defaults: &job_defaults
 anchor_for_installing_dependencies: &install_dependencies
   name: Install dependencies
   command: |
+    pip2 install --upgrade virtualenv
```
Perhaps try `python -m pip install --upgrade virtualenv`?

Commit:
test

Commit:
test

Commit:
tes

Commit:
fi

Commit:
fix

Commit:
test

brianrodri commented on scripts/build_test.py:
```diff
@@ -993,6 +993,6 @@ def mock_check_call(cmd, **unused_kwargs):
 
         with self.swap(subprocess, 'check_call', mock_check_call):
-            build.build_using_webpack()
+            build.main_using_webpack()
```
Maybe we should use a flag to use webpack instead?

lilithxxx replied:
It seems webpack is used in build.py when we pass the prod_mode flag. I am not too sure about adding a flag. /cc @seanlip @kevinlee12

brianrodri commented on scripts/common.py:
```diff
@@ -164,6 +185,103 @@ def ensure_release_scripts_folder_exists_and_is_up_to_date():
         subprocess.call(['git', 'pull', remote_alias])
 
 
+def is_port_close(port):
```
Actually, `is_port_open` is a better name (positive checks are easier to understand than negative checks), so please use that and update logic accordingly. `not s.connect_ex(('localhost', port))`

lilithxxx replied:
Done

lilithxxx replied:
Done

brianrodri commented on scripts/common.py:
```diff
@@ -164,6 +185,103 @@ def ensure_release_scripts_folder_exists_and_is_up_to_date():
+        return bool(s.connect_ex(('localhost', port)))
+
+
+# Credits: https://stackoverflow.com/a/20691431/11755830
```
Move this into the function's docstring (afterline 205)

lilithxxx replied:
Done

brianrodri commented on scripts/common.py:
```diff
@@ -164,6 +185,103 @@ def ensure_release_scripts_folder_exists_and_is_up_to_date():
+        gid: int. Group ID to be set.
+    """
+    for root, dirs, files in os.walk(path):
+        for directory in dirs:
```
nit: stay consistent: `for dir in dirs:` or `for directory in directories:`

here and elsewhere

lilithxxx replied:
Done

brianrodri commented on scripts/common.py:
```diff
@@ -164,6 +185,103 @@ def ensure_release_scripts_folder_exists_and_is_up_to_date():
+    for root, dirs, files in os.walk(path):
+        for directory in dirs:
+            os.chown(os.path.join(root, directory), uid, gid)
+        for filename in files:
```
nit: stay consistent: `for file in files:` or `for filename in filenames:`

here and elsewhere

lilithxxx replied:
Done

brianrodri commented on scripts/install_third_party_libs.py:
```diff
@@ -0,0 +1,266 @@
+    if hasattr(pip, 'main'):
+        pip.main(args=[
+            'install', '%s==%s' % (package, version), '--target', install_path])
+    # For pip version > 10.
```
Do you mean >= 10?

lilithxxx replied:
Removed this as now subprocess is used

brianrodri commented on scripts/install_third_party_libs.py:
```diff
@@ -0,0 +1,266 @@
+            'install', '%s==%s' % (package, version), '--target', install_path])
+    # For pip version > 10.
+    else:
+        #  pylint: disable=no-name-in-module, import-error
```
one too many leading spaces here and below

lilithxxx replied:
Done

brianrodri commented on scripts/install_third_party_libs.py:
```diff
@@ -0,0 +1,266 @@
+    if hasattr(pip, 'main'):
+        pip.main(args=[
+            'install', '%s==%s' % (package, version), '--target', install_path])
+    # For pip version > 10.
```
Did you mean >= 10?

lilithxxx replied:
Removed this as now subprocess is used

brianrodri commented on scripts/pylint_extensions.py:
```diff
@@ -21,14 +21,19 @@
+
+# pylint: disable=wrong-import-order
+# pylint: disable=wrong-import-position
+import astroid  # isort:skip
```
Just fix the import order.

lilithxxx replied:
Done

Commit:
fix conflicts

Commit:
fix

Commit:
Merge branch 'develop' into convert_bash_1

Commit:
fix

Commit:
fix test

seanlip commented on scripts/common.py:
```diff
@@ -15,12 +15,33 @@
+NODE_PATH = os.path.join(OPPIA_TOOLS_DIR, 'node-10.15.3')
+FRONTEND_DIR = 'core/templates/dev/head'
+NPM_PATH = os.path.join(NODE_PATH, 'bin/npm')
+os.environ['PATH'] = '%s/bin:' % NODE_PATH + os.environ['PATH']
```
Shouldn't you be doing sys.path.insert here, same as line 29?

lilithxxx replied:
Actually this is the node path which is required by the node modules so sys.path won't work here. This is similar to: https://github.com/oppia/oppia/blob/9f127abf148684a9c410d11c6c333f4c5fb9d9b2/scripts/pre_commit_linter.py#L1409
and:
https://github.com/oppia/oppia/blob/9f127abf148684a9c410d11c6c333f4c5fb9d9b2/scripts/pre_commit_linter.py#L1409

seanlip replied:
Ah ok. Perhaps leave a comment explaining.

Note that you've linked to two copies of the same thing, if I'm reading it right...

lilithxxx replied:
Oops I meant: 
https://github.com/oppia/oppia/blob/ae11e9dbeac66b5faeb68b7a2b1f16ccfdbc5ba6/scripts/typescript_checks.py#L45

lilithxxx replied:
Done

seanlip commented on scripts/common.py:
```diff
@@ -164,6 +185,103 @@ def ensure_release_scripts_folder_exists_and_is_up_to_date():
+            os.chmod(os.path.join(root, filename), mode)
+
+
+def print_string_after_two_new_lines(strings):
```
The function name talks about a singular "string" but is passed in a list of strings.

lilithxxx replied:
Done

seanlip commented on scripts/common.py:
```diff
@@ -164,6 +185,103 @@ def ensure_release_scripts_folder_exists_and_is_up_to_date():
+
+
+def print_string_after_two_new_lines(strings):
+    """Prints each string after two new lines.
```
Perhaps: "Prints the given strings, separating adjacent strings with two newlines."

lilithxxx replied:
Done

brianrodri commented on scripts/run_backend_tests.py:
```diff
@@ -0,0 +1,90 @@
+    build.main()
+
+    # Compile typescript files.
+    python_utils.PRINT('Compiling typescript...')
```
Didn't @ankita240796 make this optional? Please don't forget to respect that.

lilithxxx replied:
I don't see that here: https://github.com/oppia/oppia/blob/develop/scripts/run_backend_tests.sh and https://github.com/oppia/oppia/blob/develop/scripts/backend_tests.py

seanlip commented on scripts/common.py:
```diff
@@ -164,6 +185,103 @@ def ensure_release_scripts_folder_exists_and_is_up_to_date():
+    """
+    python_utils.PRINT(
+        'Checking whether %s is installed in %s' % (library, path))
+    if not os.path.exists('node_modules/%s' % library):
```
Use os.path.join here

lilithxxx replied:
Done

seanlip commented on scripts/common.py:
```diff
@@ -164,6 +185,103 @@ def ensure_release_scripts_folder_exists_and_is_up_to_date():
+        python_utils.PRINT('%s\n' % string)
+
+
+def install_npm_library(library, version, path):
```
library --> library_name or just name

lilithxxx replied:
Done

seanlip commented on scripts/create_expression_parser.py:
```diff
@@ -0,0 +1,67 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""It produces the expression parser."""
```
It --> This script

lilithxxx replied:
Done

seanlip commented on scripts/create_expression_parser.py:
```diff
@@ -0,0 +1,67 @@
+        expression_parser_definition, expression_parser_js])
+
+    for line in fileinput.input(files=[expression_parser_js], inplace=True):
+        # Inside this loop the STDOUT will be redirected to the file.
```
Which file is meant by "the file"? Please elaborate.

lilithxxx replied:
Done

seanlip commented on scripts/create_expression_parser.py:
```diff
@@ -0,0 +1,67 @@
+                line), end='')
+
+    for line in fileinput.input(files=[expression_parser_js], inplace=True):
+        # Inside this loop the STDOUT will be redirected to the file.
```
Ditto.

lilithxxx replied:
Done

Commit:
fix

seanlip commented on scripts/install_third_party_libs.py:
```diff
@@ -0,0 +1,266 @@
+        common.print_string_after_two_new_lines([
+            'Pip is required to install Oppia dependencies, but pip wasn\'t '
+            'found',
+            'on your local machine.',
```
I am a bit confused about why you want to separate the previous two strings by two newlines...?

lilithxxx replied:
Added in the same line

seanlip commented on scripts/install_third_party_libs.py:
```diff
@@ -0,0 +1,266 @@
+                'Windows%29')
+        sys.exit(1)
+
+    # For pip version < 10.
```
Shouldn't we be specifying an exact pip version that developers should install?

lilithxxx replied:
We should use the latest version of pip I think but circleCI uses pip version 9.

seanlip replied:
OK, perhaps leave a comment here.

Should we configure CircleCI to use a later version of pip? (Optional, though, and I think it's out of scope for the project. Only mentioning in case it's an easy option.)

lilithxxx replied:
Removed this as now subprocess is used

seanlip commented on scripts/install_third_party_libs.py:
```diff
@@ -0,0 +1,266 @@
+    # For pip version > 10.
+    else:
+        #  pylint: disable=no-name-in-module, import-error
+        import pip._internal
```
This is quite surprising; is there no public way to install via newer versions of pip?

lilithxxx replied:
No the other option is installing via subprocess.

seanlip replied:
Is that the "canonical" way? If so, then perhaps go ahead and install via subprocess, since that avoids referencing internal methods (which I think is not how pip is intended to be used).

lilithxxx replied:
Removed this as now subprocess is used

seanlip commented on scripts/install_third_party_libs.py:
```diff
@@ -0,0 +1,266 @@
+            for line in fileinput.input(
+                    files=[os.path.join(
+                        common.OPPIA_TOOLS_DIR,
+                        'skulpt-0.10.0/skulpt/skulpt.py')], inplace=True):
```
Here and elsewhere, break this into separate bits; don't include any directory separators in your strings. Please do this anywhere else you refer to path names.

If you need to reference this path multiple times, create a new variable for it and make use of that. Do similar things in other files too.

lilithxxx replied:
Done

seanlip commented on scripts/pre_push_hook.py:
```diff
@@ -403,8 +403,8 @@ def main():
                 python_utils.PRINT(
                     'Push aborted due to failing frontend tests.')
                 sys.exit(1)
```
I think @ankita240796 is doing a "raise Exception" in the scripts she's writing when there is an error.

I wonder if we should be consistent about this (i.e. the behaviour when there's an error) throughout our scripts? It might be worth coordinating with her on this.

@brianrodri -- do you have any suggestions on what the preferred behaviour should be? sys.exit(1) or raise Exception?

brianrodri replied:
Both work fine (returns a value of 1 to bash). I'd lean towards raising an Exception since we can provide more helpful details about why things failed.

seanlip replied:
SGTM. @lilithxxx please take note (if you agree)!

lilithxxx replied:
Done

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,211 @@
+    while not common.is_port_close(4444) or not common.is_port_close(9001):
+        time.sleep(1)
+
+    if os.path.isdir('../protractor-screenshots'):
```
Split paths and don't include dir separators in the strings. Ditto below and elsewhere.

lilithxxx replied:
Done

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,211 @@
+
+def cleanup():
+    """Send a kill signal to the dev server and Selenium server."""
+    common.kill_process(4444)
```
Pull the port numbers out into top-of-file constants.

lilithxxx replied:
Done

seanlip commented on scripts/run_frontend_tests.py:
```diff
@@ -0,0 +1,88 @@
+    setup.main()
+    setup_gae.main()
+    xvfb_prefix = ''
+    if os.environ.get('VAGRANT') or os.path.isfile('/etc/is_vagrant_vm'):
```
Note #7521 -- Vagrant is being removed due to lack of usage. I suspect that PR will get merged before this one, so you might want to consider dropping the Vagrant stuff.

lilithxxx replied:
Done

seanlip commented on scripts/run_frontend_tests.py:
```diff
@@ -0,0 +1,88 @@
+
+    if xvfb_prefix:
+        subprocess.call([
+            xvfb_prefix, 'node_modules/karma/bin/karma', 'start',
```
Use os.path.join; use variables or constants if you're using the same path in multiple places.

lilithxxx replied:
Done

seanlip commented on scripts/run_presubmit_checks.py:
```diff
@@ -0,0 +1,92 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""It runs the following tests in all cases.
```
Here and elsewhere: "It" --> "This script" when you are starting off a file overview. Otherwise no one knows what "it" refers to.

lilithxxx replied:
Done

seanlip reviewed (changes requested):
Thanks @lilithxxx, took a quick initial pass.

Commit:
fix conflicts

Commit:
fix

Commit:
address comments

Commit:
fix

Commit:
fix node

Commit:
fix

Commit:
remove vagrant

Commit:
fix

Commit:
address comments

Commit:
fix conflicts

Commit:
fix lint

lilithxxx commented:
@seanlip @brianrodri PTAL

Commit:
fix

Commit:
fix

seanlip commented on scripts/common.py:
```diff
@@ -165,6 +188,104 @@ def ensure_release_scripts_folder_exists_and_is_up_to_date():
+                process.send_signal(signal.SIGTERM)
+
+
+def run_command(command):
```
Let's delete this function, and just pass stuff to subprocess directly using lists.

lilithxxx replied:
Done

seanlip reviewed (approved):
Thanks, @lilithxxx! Just one small note, but I didn't see any other issues.

seanlip commented:
@brianrodri @kevinlee12 @nithusha21 PTAL for a second round of reviews.

Commit:
address comments

Commit:
fix

Commit:
fix

Commit:
fix

Commit:
fix

seanlip commented:
Hi @lilithxxx, I think Travis CI is still failing...

seanlip commented:
You might need to ensure that your Python process has edit access to the relevant directories. There are a number of suggested solutions if you search, e.g. https://github.com/Polymer/polymer-starter-kit/issues/54#issuecomment-106941727

Commit:
fix

brianrodri commented on scripts/run_frontend_tests.py:
```diff
@@ -0,0 +1,78 @@
+        'on your filesystem.',
+        'Running test in development environment'])
+
+    build.main()
```
All `main` functions should call others by explicitly passing in a list of arguments. This should allow us to get use `parse_args` instead.

```
build.main(argv=[])
```

lilithxxx replied:
Done

brianrodri commented on scripts/run_backend_tests.py:
```diff
@@ -0,0 +1,93 @@
+from . import setup_gae
+
+_PARSER = argparse.ArgumentParser()
+_PARSER.add_argument(
```
Where are the other flags?

lilithxxx replied:
What other flags are we talking about?

brianrodri replied:
```bash
# (1) Generate a coverage report by adding the argument	
#	
#   --generate_coverage_report	
#	
# but note that this will slow down the tests by a factor of 1.5 or more.	
#	
# (2) Append a test target to make the script run all tests in a given module	
# or class, or run a particular test. For example, appending	
#	
#   --test_target='foo.bar.Baz'	
#	
# runs all tests in test class Baz in the foo/bar.py module, and appending	
#	
#   --test_target='foo.bar.Baz.quux'	
#	
# runs the test method quux in the test class Baz in the foo/bar.py module.	
#	
# (3) Append a test path to make the script run all tests in a given	
# subdirectory. For example, appending	
#	
#   --test_path='core/controllers'	
#	
# runs all tests in the core/controllers/ directory.	
#	
# (4) Enable the verbose log by add the argument. It will display the outputs of	
#   the tests being run.	
#	
#   --verbose or -v
```

lilithxxx replied:
Oh they're all still present in scripts/backend_tests.py. This file is run_backend_tests.py

brianrodri replied:
Hmm that's really confusing, can we just expand and rename that file instead of introducing an indirection?

lilithxxx replied:
Done

brianrodri commented on scripts/run_e2e_tests.py:
```diff
@@ -0,0 +1,243 @@
+from . import setup
+from . import setup_gae
+
+_PARSER = argparse.ArgumentParser()
```
Pass a `description` argument with the helpful information kept in the `INSTRUCTIONS` sections of the old bash scripts for each migrated script.

```python
_PARSER = argparse.ArgumentParser(description="""
Run this script from the oppia root folder:	

    python -m scripts.run_e2e_tests

The root folder MUST be named 'oppia'.

Note: You can replace 'it' with 'fit' or 'describe' with 'fdescribe' to
run a single test or test suite.
""".strip())
```

lilithxxx replied:
Done

Commit:
merge develop

Commit:
address comments

Commit:
fix

Commit:
fix

Commit:
fix

Commit:
only one backend_tests.py

Commit:
address comments

Commit:
Merge branch 'develop' into convert_bash_1

Commit:
fix

Commit:
fix

Commit:
fix

brianrodri commented on scripts/build.py:
```diff
@@ -1393,4 +1394,4 @@ def build():
 # it will only be called when build.py is used as a script.
 if __name__ == '__main__':  # pragma: no cover
-    build()
+    main(argv=sys.argv[1:])
```
Why skip the first argument (program name)? `parse_args` should work correctly even with it there.

lilithxxx replied:
No its not working then. Its saying unrecognised args: 'file_name.py'

brianrodri replied:
It works fine locally (passing None as the value of argv), sounds like you're making an extra call? Could you provide a stack trace?

lilithxxx replied:
yes it works fine with none but not with only sys.argv. Its reading an extra argument(the filename). I guess this is expected?

brianrodri replied:
That's right. Do not pass `sys.argv`, pass `None` instead (that's why I asked the default value to be `None`).

lilithxxx replied:
Wait now I am confused. If we call this script (build.py) directly from terminal with some args then the main function will get executed. But if the we are passing None to main always, then those args are never gonna get picked up, right? Instead, I think we should pass sys.argv to main here and in the definition of main, we should pass None. Something like:
```python
def main(argv=None):
    # Code.

if __name__ == '__main__':
    main(argv=sys.argv[1:])
```

brianrodri replied:
No, we are passing `None` because that is the default value `parse_args` expects. The [docs](https://docs.python.org/2/library/argparse.html#argparse.ArgumentParser.parse_args) for it explain:
> [args](https://docs.python.org/2/library/argparse.html#args) - List of strings to parse. The default is taken from [`sys.argv`](https://docs.python.org/2/library/sys.html#sys.argv).

So the flow is this.

```python
if __name__ == '__main__':
    # argv will be None
    main()

def main(argv=None):
    # args is None as well, sys.argv will be checked.
    parser.parse_args(args=argv)

    # We explicitly pass empty list because we don't want
    # sys.argv to be used.
    other_module.main(args=[])
```

Passing `sys.argv` directly does the wrong thing as you've already noticed, so we pass `None` instead so that `parse_args` can do the right thing.

brianrodri replied:
As a side-note, let's replace `argv` with `args` to make this less confusing.

lilithxxx replied:
Done

brianrodri commented on scripts/build.py:
```diff
@@ -1349,22 +1350,22 @@ def compile_typescript_files_continuously(project_dir):
     """
-    parser = optparse.OptionParser()
-    parser.add_option(
+    parser = argparse.ArgumentParser()
```
Please stick to your convention of defining this as a top-level constant.

lilithxxx replied:
Done

brianrodri commented on .circleci/config.yml:
```diff
@@ -15,8 +15,8 @@ anchor_for_job_defaults: &job_defaults
   command: |
-    source scripts/setup.sh || exit 1
-    source scripts/setup_gae.sh || exit 1
+    sudo pip install pyyaml
```
Please get `python -m scripts.setup` to handle this as discussed.

lilithxxx replied:
I am getting errors even after installing locally. This and the following two imports are required to be installed as sudo to run the tests in circleci. Also, it was here from before.

brianrodri replied:
Is this resolved? `install_third_party_libraries` should now take care of this, right?

brianrodri replied:
Still waiting here.

lilithxxx replied:
Sorry this got lost in the long list of reviews. Done!

brianrodri commented on .circleci/config.yml:
```diff
@@ -101,14 +100,18 @@ jobs:
       - restore_cache:
           <<: *restore_cache
-      - run: sudo pip install pyyaml
+      - run: sudo pip install webtest
```
Call `scripts.setup` instead.

lilithxxx replied:
Ditto

brianrodri commented on scripts/create_expression_parser.py:
```diff
@@ -0,0 +1,72 @@
+        'ExpressionParserService.js')
+
+    # Install the basic environment, e.g. nodejs.
+    install_third_party_libs.main()
```
Explicitly pass empty args.

brianrodri commented on scripts/create_expression_parser.py:
```diff
@@ -0,0 +1,72 @@
+
+def main():
+    """Produces the expression parser."""
+    setup.main()
```
Explicitly pass empty args.

lilithxxx replied:
Setup and setup_gae accepts no arguments and hence its main accepts no argv

brianrodri commented on scripts/install_third_party_libs.py:
```diff
@@ -0,0 +1,271 @@
+    if run_minified_tests is True:
+        python_utils.PRINT(
+            'Running build task with concatenation and minification')
+        build.main(argv=['--prod_env'])
```
Are you sure we should always run in prod mode?

lilithxxx replied:
That's how it is in the original script.

brianrodri commented on scripts/install_third_party_libs.py:
```diff
@@ -0,0 +1,271 @@
+def main(argv=None):
+    """Install third-party libraries for Oppia."""
+    setup.main()
+    setup_gae.main()
```
Explicitly pass empty args.

lilithxxx replied:
Setup and setup_gae accepts no arguments and hence its main accepts no argv

brianrodri commented on scripts/install_third_party_libs.py:
```diff
@@ -0,0 +1,271 @@
+
+    # Download and install required JS and zip files.
+    python_utils.PRINT('Installing third-party JS libraries and zip files.')
+    install_third_party.main()
```
Explicitly pass empty args

lilithxxx replied:
install_third_party accepts no arguments

brianrodri commented on scripts/run_backend_tests.py:
```diff
@@ -308,8 +305,14 @@ def _get_test_target_classes(path):
+def main(argv=None):
     """Run the tests."""
+    setup.main()
+    setup_gae.main()
```
Explicitly pass empty args

lilithxxx replied:
Setup and setup_gae accepts no arguments and hence its main accepts no argv

brianrodri commented on scripts/run_backend_tests.py:
```diff
@@ -81,7 +80,11 @@
+_PARSER = argparse.ArgumentParser(description="""
+    Run this script from the oppia root folder:
+
+        python -m scripts.run_backend_tests""")
```
Missing details about how test_path/test_target are mutually exclusive?

lilithxxx replied:
Done

brianrodri commented on scripts/run_backend_tests.py:
```diff
@@ -81,7 +80,11 @@
+
+        python -m scripts.run_backend_tests""")
+
 _PARSER.add_argument(
```
Use a [mutually exclusive group](https://docs.python.org/2/library/argparse.html#argparse.ArgumentParser.add_mutually_exclusive_group) to force test_path/test_target to be mutually exclusive.

lilithxxx replied:
Done

Commit:
fix

brianrodri commented on scripts/run_tests.py:
```diff
@@ -0,0 +1,61 @@
+def main():
+    """Run all the tests."""
+    setup.main()
+    setup_gae.main()
```
Explicitly pass empty args

lilithxxx replied:
Setup and setup_gae accepts no arguments and hence its main accepts no argv

brianrodri replied:
See below.

brianrodri commented on scripts/start.py:
```diff
@@ -0,0 +1,209 @@
+def main(argv=None):
+    """Starts up a development server running Oppia."""
+    setup.main()
+    setup_gae.main()
```
Explicitly pass empty args

lilithxxx replied:
Setup and setup_gae accepts no arguments and hence its main accepts no argv

brianrodri replied:
That's why you should be passing an empty argument list. This is currently accepting `sys.argv` which will be an array with the filename at the front.

This (and every other call to main outside of `if __name__ == '__main__'`) should look like: `module.main(argv=[])`.

lilithxxx replied:
I meant setup.py  and setup_gae.py does not accept any arguments. There is no argparse and parse_args in setup.py. Its never run with any arguments.

brianrodri replied:
I see, in that case add argparse into them anyway for consistency. You can move the script instructions/description into the argparse object so that it's not totally useless (it'll be nice to have `-h` for printing help).

lilithxxx replied:
Hmm then we will be getting pylint errors that the variable _PARSER is never used. I think we can add them later if any flags are introduced in the setup file

brianrodri replied:
Why would it be unused? Just call `parse_args` like you would any other parser, and ignore the return value.

lilithxxx replied:
Done

Commit:
fix errors

Commit:
address comments

Commit:
fix

Commit:
use None and args

brianrodri commented on scripts/setup.py:
```diff
@@ -0,0 +1,182 @@
+"""
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
```
Do necessary pip installs here.

brianrodri commented on scripts/start.py:
```diff
@@ -0,0 +1,204 @@
+
+
+if __name__ == '__main__':
+    main(args=None)
```
Just leave this as `main()` here and in the other `if __name__ == '__main__'` blocks.

lilithxxx replied:
Done

Commit:
fix permission

Commit:
fix

Commit:
fix conflicts

brianrodri commented on scripts/create_expression_parser.py:
```diff
@@ -0,0 +1,68 @@
+from . import setup
+
+
+def main():
```
```python
def main(args=None):
    unused_parsed_args = _PARSER.parse_args(args)
```

lilithxxx replied:
Done

Commit:
address comments

lilithxxx commented:
@brianrodri I have addressed your comments. PTAL!

brianrodri commented on scripts/create_expression_parser.py:
```diff
@@ -0,0 +1,76 @@
+def main(args=None):
+    """Produces the expression parser."""
+    unused_parsed_args = _PARSER.parse_args(args=args)
+    setup.main(args=[])
```
`main(args=args)`

brianrodri commented on scripts/install_third_party_libs.py:
```diff
@@ -0,0 +1,272 @@
+
+
+def install_skulpt(args):
+    """Download and install Skulpt. Skulpt is built using a Python script
```
Move this into `_PARSER`'s description

lilithxxx replied:
This is a function docstring not the overview of what the script does, so shouldn't is just stay here?

brianrodri replied:
True, ok fine with me.

brianrodri commented on scripts/install_third_party_libs.py:
```diff
@@ -0,0 +1,272 @@
+    # 374076889.
+    subprocess.call([common.NPM_PATH, 'dedupe'])
+
+    install_skulpt(args)
```
Let's parse the args in `main` and pass the parsed data into `install_skulpt`.

lilithxxx replied:
Done

brianrodri commented on scripts/run_presubmit_checks.py:
```diff
@@ -0,0 +1,105 @@
+# limitations under the License.
+
+"""This script runs the following tests in all cases.
+- Javascript and Python Linting
```
Move this information into `_PARSER`'s description.

lilithxxx replied:
Done

brianrodri commented on scripts/setup_gae.py:
```diff
@@ -0,0 +1,106 @@
+
+_PARSER = argparse.ArgumentParser(description="""
+    This file should not be invoked directly, but called from other Python
+    scripts.
```
So why don't we make this a module-function or move it to `common.py`? That way, it becomes impossible to invoke directly at all.

lilithxxx replied:
I just copied the description of the original file here. Thinking about it now, I don't see what's the harm in calling it directly. It can be called directly just to set up the environment I guess. I removed this line from the description

brianrodri commented on scripts/build.py:
```diff
@@ -1349,22 +1358,14 @@ def compile_typescript_files_continuously(project_dir):
+def main(args=None):
     """The main method of this script.
 
     Creates a third-party directory where all the JS and CSS dependencies are
```
Move this into `_PARSER`'s description.

lilithxxx replied:
Done

brianrodri reviewed (changes requested):
Awesome, getting really close! Here are a few general requests:
- For all `main()` functions, change `_PARSER.parse_args` to be the very first statement called. Right now, many scripts will perform complicated work even when a caller only wants to print the help message with `-h` (which internally calls `sys.exit(0)` for us).
- Remove the leading indentation in the `_PARSER` description strings -- it'll look weird on the command-line. Just do something like this (disabling pylint if necessary):
   ```python
   argparse.ArgumentParser(description="""
   No leading indentation, just a block of text;
   possibly with many lines.
   It's fine to indent some lines, just not all of them. For example:
       python -m this_module -h
   """)
   ```
- Double-check we are only calling `main()` with no args from the `if __name__ == '__main__'` block. **Every other call should explicitly pass a list,** otherwise `sys.argv` will cause subtle errors (we might want to add a lint check in the future).

brianrodri commented on scripts/install_third_party_libs.py:
```diff
@@ -0,0 +1,272 @@
+    warning saying its dist command will not work properly without GitPython,
+    but it does actually work due to the patches.
+    """
+    parsed_args = _PARSER.parse_args(args=args)
```
Move this into main and pass the `parsed_args` as a parameter to `install_skulpt` instead.

lilithxxx replied:
This is done, right?

Commit:
fix e2e

seanlip commented:
This is interesting -- @lilithxxx, the e2e tests are now running into this error:

``` 
08:31:20.232 ERROR [BaseServer.start] - Port 4444 is busy, please choose a free port and specify it using -port option
Exception in thread "main" java.lang.RuntimeException: java.net.BindException: Address already in use
```

Should be straightforward to fix? You can check what process is using port 4444 and/or use a different port, perhaps...

Commit:
address comments

Commit:
fix

lilithxxx commented:
@brianrodri addressed all your comments.
> Double-check we are only calling main() with no args from the if __name__ == '__main__' block. Every other call should explicitly pass a list, otherwise sys.argv will cause subtle errors 

Done

lilithxxx commented:
@seanlip I am getting a different error now: 
```
  Message:
    Failed: Angular could not be found on the page http://localhost:9001/signup?return_url=http%3A%2F%2Flocalhost%3A9001%2F. If this is not an Angular application, you may need to turn off waiting for Angular.
                              Please see 
                              https://github.com/angular/protractor/blob/master/docs/timeouts.md#waiting-for-angular-on-page-load
  Stack:
    Error: Angular could not be found on the page http://localhost:9001/signup?return_url=http%3A%2F%2Flocalhost%3A9001%2F. If this is not an Angular application, you may need to turn off waiting for Angular.
                              Please see 
                              https://github.com/angular/protractor/blob/master/docs/timeouts.md#waiting-for-angular-on-page-load
        at executeAsyncScript_.then (/home/travis/build/oppia/oppia/node_modules/protractor/built/browser.js:720:27)
        at ManagedPromise.invokeCallback_ (/home/travis/build/oppia/oppia/node_modules/selenium-webdriver/lib/promise.js:1376:14)
        at TaskQueue.execute_ (/home/travis/build/oppia/oppia/node_modules/selenium-webdriver/lib/promise.js:3084:14)
        at TaskQueue.executeNext_ (/home/travis/build/oppia/oppia/node_modules/selenium-webdriver/lib/promise.js:3067:27)
        at asyncRun (/home/travis/build/oppia/oppia/node_modules/selenium-webdriver/lib/promise.js:2927:27)
        at /home/travis/build/oppia/oppia/node_modules/selenium-webdriver/lib/promise.js:668:7
        at process._tickCallback (internal/process/next_tick.js:68:7)
    From: Task: Run beforeAll in control flow
        at UserContext.<anonymous> (/home/travis/build/oppia/oppia/node_modules/jasminewd2/index.js:94:19)
    From asynchronous test: 
    Error
        at Suite.<anonymous> (/home/travis/build/oppia/oppia/core/tests/protractor_desktop/collections.js:40:3)
        at Object.<anonymous> (/home/travis/build/oppia/oppia/core/tests/protractor_desktop/collections.js:30:1)
        at Module._compile (internal/modules/cjs/loader.js:701:30)
        at Object.Module._extensions..js (internal/modules/cjs/loader.js:712:10)
        at Module.load (internal/modules/cjs/loader.js:600:32)
        at tryModuleLoad (internal/modules/cjs/loader.js:539:12)
```
Any idea why this is happening?

Commit:
fix

Commit:
skip e2e

brianrodri reviewed (commented):
Still one comment you haven't addressed, otherwise LGTM

brianrodri commented on .circleci/config.yml:
```diff
@@ -101,14 +99,18 @@ jobs:
-      - run: sudo pip install pyyaml
+      - run: sudo pip install webtest
+      - run: sudo pip install configparser
+      - run: sudo pip install coverage
```
Do we still need these as well? Shouldn't `install_third_party_libs` take care of them?

lilithxxx replied:
Yes these are required else there are some errors in CircleCI related to environment. These are also installed in the install_third_party_libs but still they are needed here

brianrodri replied:
If every script can run without it, that's fine with me.

Commit:
remove yaml

lilithxxx commented:
@kevinlee12 @nithusha21 PTAL as code-owners.

Commit:
fix lint

seanlip commented:
@DubeySandeep @vojtechjelinek @aks681 PTAL too. Thanks!

nithusha21 commented on scripts/install_chrome_on_travis.py:
```diff
@@ -0,0 +1,61 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This script should only be ran by Travis to install and provide a constant
```
```suggestion
"""This script should only be run by Travis to install and provide a constant
```

lilithxxx replied:
Done

nithusha21 commented on scripts/install_chrome_on_travis.py:
```diff
@@ -0,0 +1,61 @@
+import python_utils
+
+_PARSER = argparse.ArgumentParser(description="""
+This script should only be ran by Travis to install and provide a constant
```
```suggestion
This script should only be run by Travis to install and provide a constant
```

nithusha21 replied:
Also, shouldn't these lines ideally be indented by two? I think you can use single quote strings instead of """ as we typically use the latter for docstrings.

lilithxxx replied:
Done and used single quotes

brianrodri replied:
The description is exceptional since it typically has a ton of information to give. Also, indents show up in the command line and makes the output look weird.

I'm fine with this case since it's just one sentence, but I think it'd be a better idea to use triple-quotes for >1 paragraph descriptions.

lilithxxx replied:
I tested with single quotes and the indent does not get messed up. So should I still revert the last commit?

lilithxxx replied:
Also isn't it better to have a use only one type of quotes throughout all descriptions, just to maintain consistency

brianrodri replied:
I agree, having the description break at 80-characters is a nice plus too for very long sentences (which will not happen here).

I vote for triple-quote strings and no indentations, what do you think @nithusha21?

nithusha21 reviewed (approved):
Overall LGTM! Just some minor nits. Great work btw!

Commit:
use single quotes

kevinlee12 reviewed (approved):
lgtm!

DubeySandeep reviewed (approved):
LGTM for the code owners file: `.github/CODEOWNERS`

vojtechjelinek reviewed (approved):
LGTM from code owner's perspective. Thanks!

lilithxxx commented:
@aks681 PTAL?

aks681 reviewed (approved):
lgtm as codeowner

aks681 commented:
@lilithxxx Note merge conflicts

Commit:
fix conflicts

Commit:
fix conflicts

Commit:
test

Commit:
fix linr

brianrodri commented:
Why are there so many newly open files?

brianrodri commented on scripts/run_performance_tests.py:
```diff
@@ -0,0 +1,165 @@
+from . import setup_gae
+
+_PARSER = argparse.ArgumentParser(
+    description='The root folder MUST be named \'oppia\'. '
```
Yeah I really don't like these after all, this will just print a huge blob of text. Please use triple quote strings instead and add proper newlines/indentations/etc for nice formatting.

lilithxxx replied:
Done

Commit:
add tests

Commit:
use triple strings

vojtechjelinek commented:
@lilithxxx Hi, what is the reason behind not migrating the `run_e2e_tests.sh` to Python?

lilithxxx commented:
@vojtechjelinek this is being done in https://github.com/oppia/oppia/pull/7598
